cgroup v2 helpers

rdna opened the pull request:
Add a bunch of cgroup v2 helpers and use them in examples.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,77 @@
+]
+
+
+def sock_cgroup(skcd):
```
The kernel has `sock_cgroup_ptr`, how about we make this return a `struct cgroup *` instead of `struct cgroup` and rename it for consistency?

rdna replied:
Lol, that what I initially did, but then I decided to make it "more python-friendly". Sure, can convert it back to `_ptr` version.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,77 @@
+
+    :return: ``struct cgroup`` object.
+    """
+    return Object(skcd.prog_, 'struct cgroup', address=skcd.val)
```
This can be `cast('struct cgroup *', skcd.val)`.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,77 @@
+    return None
+
+
+def cgroup_name(cgrp):
```
sysfs also uses kernfs, so it'd be worth it to add `kernfs_name` and `kernfs_path` helpers and use those for `cgroup_name` and `cgroup_path`. Would you mind doing that?

rdna replied:
`kernfs_name` is simple, I'll do it

`kernfs_path` is more involved so I'll do simplified version of it, that later can be extended if needed. By "simplified" I mean I won't implement the whole logic of `kernfs_path_from_node_locked` that is used by `kernfs_path` underneath but will handle only scenario when `from == NULL`, i.e. the path is built from starting root, not from a random `kn` in the fs. LMK if it's not a good idea for some reason.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,77 @@
+        names.append(cgroup_name(cgrp))
+        cgrp = cgroup_parent(cgrp)
+    names.reverse()
+    return os.path.join(b'', *names)
```
This is really pedantic, but I'd prefer `b'/'.join(names)` over `os.path.join()`, because we're joining a Linux path, not a path on whatever operating system drgn is running on. And who knows, maybe someone will port drgn to Windows and use it to debug a Linux core dump :)

rdna replied:
Yeah, I saw it in other helpers, but wasn't sure "why" :) OK, makes sense, I'll change it -- not a big deal.

rdna commented:
In v2:
- renamed `sock_cgroup` to `sock_cgroup_ptr` and changed return type to ptr;
- use `cast` instead of `Object` in `sock_cgroup_ptr`;
- implement `kernfs` helpers and use them in `cgroup_path`, `cgroup_name`;
- avoid `os.path.join`;
- implement additional cgroup v2 helpers to traverse hierarchy and an example -- @osandov please  lmk if it's not a good idea to add new commits into this PR, I can move them into separate one, not a big deal.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,183 @@
+        next_ = container_of(
+            parent.children.next, 'struct cgroup_subsys_state', 'sibling'
+        )
+    elif not (pos.flags.value_() & css_flags.CSS_RELEASED):
```
You don't need the `.value_()` here, drgn objects support all of the usual operators. This can be changed in a few other places, too.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,183 @@
+            parent.children.address_of_(),
+            'sibling'
+        ):
+            if next_.serial_nr.value_() > pos.serial_nr.value_():
```
Likewise here, you don't need `.value_()`. Although in this case, it would be a bit more efficient to read `pos.serial_nr` once (under the covers, each access needs to read from `/proc/kcore` or the core dump):

```
pos_serial_nr = pos.serial_nr.value_()
for ...:
    if next_.serial_nr > pos_serial_nr:
        break
```

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,183 @@
+
+    if next_.sibling.address_of_() != parent.children.address_of_():
+        return next_
+    return None
```
Hm, I don't know how consistent I've been with this, but I think that for helpers that return a pointer, we should return `NULL` instead of `None`. You can use `NULL('struct cgroup_subsys_state *')`

rdna commented on drgn/helpers/linux/bpf.py:
```diff
@@ -0,0 +1,76 @@
+    except KeyError:
+        enumerators = [
+            (name, value)
+            for (name, value) in prog['__MAX_BPF_ATTACH_TYPE'].type_.enumerators
```
> Since this type isn't anonymous, we can use `prog.type('enum bpf_attach_type')` instead of `prog['__MAX_BPF_ATTACH_TYPE'].type`. Also, we probably want to remove the redundant `BPF_` prefix?

Oh, `prog.type` -- that's definitely better. I tried to do it with `prog['enum bpf_attach_type']`, failed and decided to go with what worked. Sure, will switch to it.

As for `BPF_` prefix IMO it's better to keep it so that e.g. I can copy/paste it from UAPI to my drgn script. Also it simplifies auto-generating types for enums in general -- just use the name as is.

We did a different thing for TCP states enum, but that one is very common in many user-space utils and people got used to `LISTEN` more than `TCP_LISTEN` so there it makes sense.



> At this point I wonder if we want to factor this out into a utility function in `drgn/helpers/__init__.py`. Something like `get_enum(name, type, prefix)`? Feel free to suggest a better name.

Yep, I had same thoughts, but .. I was lazy :) I added the function in `drgn/helpers/__init__.py` and wen can bikesched on names in the next update.

rdna commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,183 @@
+    return kernfs_path(cgrp.kn)
+
+
+def _get_css_flags(prog):
```
Right, it's only for internal usage.

The main reason I did it is to cache the enum to avoid `read()`ing it from kern mem all the time during iteration, but I admin that I don't know the whole picture when drgn reads and when it doesn't so maybe it's premature optimization.

I'll remove it but lmk if caching makes sense here.

osandov replied:
Enums are available in the kernel debug info, so we don't have to read kernel memory for it. So in this case, caching won't help much. I should document some of the low-level details of drgn that can help people writing helpers :)

rdna commented:
In v3:
- introduce `get_enum` and switch to it to get enum types;
- use `enum bpf_attach_type` instead of an enumerator to get the type;
- yield `prog` instead of `item.prog` in `cgroup_bpf_prog_for_each_effective`;
- don't use `value_()` where not needed;
- remove `css` enum and access enumerators directly;
- switch from `None` to `NULL`;
- switch to `itertools.count()`.

osandov commented on drgn/helpers/__init__.py:
```diff
@@ -56,3 +57,25 @@ def escape_string(buffer: Iterable[int], escape_single_quote: bool = False,
                    for c in buffer)
+
+
+def get_enum(prog: object, name : str, type : str, ignore: Iterable[str] = [],
```
`prog: drgn.Program`. Also, let's make this take the type as a `drgn.Type` instead of a string. Then the caller can deal with doing `prog.type` or `prog[...].type_` as appropriate.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,169 @@
+    every ``struct sock``.
+
+
+    :return: ``struct cgroup *`` object.
```
This annotation isn't necessary since it's documented by the C signature.

osandov commented on drgn/helpers/linux/cgroup.py:
```diff
@@ -0,0 +1,169 @@
+    parent_css = cgrp.self.parent
+    if parent_css:
+        return container_of(parent_css, 'struct cgroup', 'self')
+    return None
```
Ah here's the place I was thinking of when I said I wasn't consistent with this. Let's make this return a `NULL` object, too, then we can get rid of the `:return:` annotation and instead note that it returns `NULL` if the cgroup doesn't have a parent.

rdna commented:
In v4:
- better typing in `get_enum`;
- moar `None` to `NULL` in cgroup helpers;
- cleanup `:return:` annotations in cgroup helpers documentation.

osandov commented:
I ended up taking a different approach for the enum types. I think it makes more sense to work with drgn Objects directly in the helpers and allow the example scripts to create the `IntEnum` types for their own use. I'll merge this as soon as tests pass. Thanks again!
